subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStepTest.groovy

```diff
@@ -0,0 +1,149 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps
+
+import com.google.common.collect.ImmutableSortedMap
+import org.gradle.internal.classloader.ClassLoaderHierarchyHasher
+import org.gradle.internal.execution.AfterPreviousExecutionContext
+import org.gradle.internal.execution.BeforeExecutionContext
+import org.gradle.internal.execution.CachingResult
+import org.gradle.internal.execution.Step
+import org.gradle.internal.execution.UnitOfWork
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint
+import org.gradle.internal.fingerprint.impl.AbsolutePathFingerprintingStrategy
+import org.gradle.internal.hash.HashCode
+import org.gradle.internal.snapshot.FileSystemSnapshot
+import org.gradle.internal.snapshot.ValueSnapshot
+import org.gradle.internal.snapshot.ValueSnapshotter
+import org.gradle.internal.snapshot.impl.ImplementationSnapshot
+import spock.lang.Specification
+import spock.lang.Subject
+
+@Subject(CaptureStateBeforeExecutionStep)
+class CaptureStateBeforeExecutionStepTest extends Specification {
+
+    def classloaderHierarchyHasher = Mock(ClassLoaderHierarchyHasher)
+    def valueSnapshotter = Mock(ValueSnapshotter)
+    def context = Mock(AfterPreviousExecutionContext)
+    def work = Mock(UnitOfWork)
+    def result = Mock(CachingResult)
+    Step<BeforeExecutionContext, CachingResult> delegate = Mock()
+    def implementationSnapshot = ImplementationSnapshot.of("MyWorkClass", HashCode.fromInt(1234))
+
+    def step = new CaptureStateBeforeExecutionStep(classloaderHierarchyHasher, valueSnapshotter, delegate)
+
+    def "no state is captured when task history is not maintained"() {
+        when:
+        step.execute(context)
+        then:
+        1 * context.work >> work
+        1 * work.isTaskHistoryMaintained() >> false
+        1 * delegate.execute { BeforeExecutionContext context ->
+            context.beforeExecutionState.empty
+        } >> result
+        0 * _
+    }
+
+    def "implementations are snapshotted"() {
+        def additionalImplementations = [
+            ImplementationSnapshot.of("FirstAction", HashCode.fromInt(2345)),
+            ImplementationSnapshot.of("SecondAction", HashCode.fromInt(3456))
+        ]
+
+        when:
+        step.execute(context)
+        then:
+        1 * work.visitImplementations(_) >> { UnitOfWork.ImplementationVisitor visitor ->
+            visitor.visitImplementation(implementationSnapshot)
+            additionalImplementations.each {
+                visitor.visitAdditionalImplementation(it)
+            }
+        }
+        interaction { fingerprintInputs() }
+        1 * delegate.execute { BeforeExecutionContext context ->
```

Comment:
This only works for assertions on arguments, not for behaviour, and is a new feature in Spock 1.3. That is why I can't use it in line 70 and 92 since I actually want the code to be executed there.

Reply:
When I change line 70 to this, the test still runs fine for me:

```
        1 * work.visitImplementations { UnitOfWork.ImplementationVisitor visitor ->
```

Reply:
You are right, it runs fine. Though this notation is a shortcut for
```
1 * work.visitImplementations({ visitor -> })
```

So the closure is a code argument constraint: https://github.com/spockframework/spock/blob/master/docs/release_notes.adoc#code-argument-constraints-are-treated-as-implicit-assertions

While the syntax is nicer, it does actually mean something different. So I would prefer to keep it as is.

Reply:
How is this different to `delegate.execute(_) { ... }` vs `delegate.execute { ... }` ?

Reply:
In `delegate.execute { ... }` the assertions are in the `{ }`, not some behaviour.

Reply:
I see what you mean now. So if I changed `delegate.execute { ... }` to `delegate.execute(_) { ... }` any failing assertions inside the closure would not fail the test. I think this is very error-prone.